x-pack/test/functional/page_objects/lens.ts

```diff
@@ -6,20 +6,147 @@
 
 import { FtrProviderContext } from '../ftr_provider_context';
 
-export function LensPageProvider({ getService }: FtrProviderContext) {
+export function LensPageProvider({ getService, getPageObjects }: FtrProviderContext) {
   const testSubjects = getService('testSubjects');
+  const retry = getService('retry');
+  const find = getService('find');
+  const browser = getService('browser');
+  const PageObjects = getPageObjects([
+    'header',
+    'common',
+    'visualize',
+    'dashboard',
+    'header',
+    'timePicker',
+  ]);
 
   return {
-    async openIndexPatternFiltersPopover() {
+    /**
+     * Clicks the index pattern filters toggle.
+     */
+    async toggleIndexPatternFiltersPopover() {
       await testSubjects.click('lnsIndexPatternFiltersToggle');
     },
 
+    /**
+     * Toggles the field existence checkbox.
+     */
     async toggleExistenceFilter() {
+      await this.toggleIndexPatternFiltersPopover();
       await testSubjects.click('lnsEmptyFilter');
+      await this.toggleIndexPatternFiltersPopover();
     },
 
     async findAllFields() {
       return await testSubjects.findAll('lnsFieldListPanelField');
     },
+
+    /**
+     * Move the date filter to the specified time range, defaults to
+     * a range that has data in our dataset.
+     */
+    goToTimeRange(fromTime = '2015-09-19 06:31:44.000', toTime = '2015-09-23 18:31:44.000') {
+      return PageObjects.timePicker.setAbsoluteRange(fromTime, toTime);
+    },
+
+    /**
+     * Wait for the specified element to have text that passes the specified test.
+     *
+     * @param selector - the element selector
+     * @param test - the test function to run on the element's text
+     */
+    async assertExpectedText(selector: string, test: (value?: string) => boolean) {
+      let actualText: string | undefined;
+
+      await retry.waitForWithTimeout('assertExpectedText', 1000, async () => {
+        actualText = await find.byCssSelector(selector).then(el => el.getVisibleText());
+        return test(actualText);
+      });
```

Comment:
This looks like an infinite loop- could you add a max # of retries?

Reply:
I don't think it's infinite. My reading of `waitForWithTimeout` is that it times out and throws an error when it times out. Am I wrong? If so, I don't see a way to specify # retries other than manually implementing it, which seems wrong. Should we add a max # retries to the retry service function signature?